### ISSUE TITLE
[templates] Maybe each {{ ~ }} should be its own text node?

### ISSUE THREAD

**Issue by domenic** [WICG/webcomponents#683]
> When a single text node contains multiple {{ ~ }}, these nodes previousSibling and nextSibling may refer to nodes in the preceding or the succeeding part.

If we just made each one its own text node, that would make replacement a lot simpler to reason about, and make previousSibling/nextSibling fairly intuitive, IMO.

I guess this would be a HTML parser change, though. Just a small one......

**Comment by rniwa** [comment]
That conversion already happens in `createInstance`. However, we don't keep `{{~}}` in an instance except when the text node is fully templatized and its parent is the template instance itself.

**Comment by annevk** [comment]
If you make them their own text node, would you not be able to create these by setting `textContent`? How would this syntax work in attribute values? You'd run into quite a few tricky things I think.

**Comment by rniwa** [comment]
I think we're talking about just non-attribute text nodes cases. For attributes, see https://github.com/w3c/webcomponents/blob/gh-pages/proposals/Template-Instantiation.md#41-attribute-value-setter and https://github.com/w3c/webcomponents/blob/gh-pages/proposals/Template-Instantiation.md#45-attributetemplatepart-interface

**Comment by annevk** [comment]
Given that I think it would make the most sense if we defined this in terms of https://dom.spec.whatwg.org/#contiguous-text-nodes for text nodes. Doing this on a singular text node level would be quite surprising. If it would simplify things a lot it could of course be done, but we'd have to be very careful.

**Comment by rniwa** [comment]
https://github.com/w3c/webcomponents/blob/gh-pages/proposals/Template-Instantiation.md#46-nodetemplatepart-interface pretty much does that because we'd split text nodes inside `createInstance` (see https://github.com/w3c/webcomponents/blob/gh-pages/proposals/Template-Instantiation.md#43-creating-template-parts) but we'd re-apply all the replacement nodes whenever any of the node template parts which belong to the same text node is updated.

I guess it's technically possible for a template content to have a contiguous text nodes so I should have devised the algorithm to treat them as one text node in the algorithm for `createInstance`. I wonder if there is any value in keeping the original split? If we had two text noes next to each other, we could either normalize them before splitting them again, or keep them separate.

**Comment by annevk** [comment]
Currently we don't have anything that distinguishes between `append(new Text("ab"));` and `append(new Text("a"), new Text("b"));`. I'm not sure if that's worth breaking for this.

**Comment by rniwa** [comment]
Yeah, that's why I went with the current approach but it's possible for authors to manually insert more Text noes into template element's content after the content had been parsed, right?

The question is whether we should be normalizing them prior to running the algorithm to split text nodes by `{{~}}` or not.

It seems like you'd prefer always normalizing first? FWIW, I'm fine with that.

**Comment by annevk** [comment]
Yeah, I'd prefer that if you have multiple Text/CDATASection nodes in a row that end up spelling {{~}} that should still work. (Note that this is more than what `normalize()` does. For some legacy reason that does not consider CDATASection nodes.) If we don't do that we add some new meaning to Text node boundaries that previously was not there and it seems nice to avoid that if possible.

**Comment by rniwa** [comment]
How about the final DOM tree's state after the substitution had occurred? i.e. after each node template part's value had been assigned.

Normalizing text nodes in that case would result in some weird edge cases like some Text nodes in a node template part not being in the final DOM tree due to them getting merged together, or even some Text nodes end up having extra contents.

If we're going to that path, it might be actually better to use something like static range to represent each part instead of having a node list.

Also see https://github.com/w3c/webcomponents/issues/680.

**Comment by rniwa** [comment]
Specifically, if we created two node template parts `a` and `b` for `{{a}} b {{c}}`, should `a.replacementNodes[0]` be identical to `a.replacementNodes[1]`? If that's the desired outcome, it's probably better to use range-like objects than keeping node lists.

**Comment by annevk** [comment]
An alternative might be that after initialization we use special Text nodes for the template bits. However, that would not be consistent with attributes. Therefore, using ranges for both might not be too bad a plan, but I believe that's not what Ember et al have done.

**Comment by rniwa** [comment]
Ember uses an empty text node for performance reasons if I remember correctly.

**Comment by justinfagnani** [comment]
> Normalizing text nodes in that case would result in some weird edge cases like some Text nodes in a node template part not being in the final DOM tree due to them getting merged together, or even some Text nodes end up having extra contents.

lit-html uses comments to demark expressions, then removes the comment nodes from the template. This causes text spans like `a {{b}} c` to produce two adjacent Text nodes, `"a "` and `" c"`. And it's true that normalization of the template or template instance would cause severe problems.

In some research it looked like some versions of IE or Edge (and not other engines) might spontaneously normalize Nodes, but wouldn't merge Text nodes that had outstanding references to them. If that's the case, could that behavior be spec'ed?

> Specifically, if we created two node template parts `a` and `b` for `{{a}} b {{c}}`, should a.replacementNodes[0] be identical to a.replacementNodes[1]

I think that `{{a}} b {{c}}` really needs to create two node template parts because we need to be able to set values to non-text values, especially other TemplateInstances for nested templating.

This is a common way to compose templates in JSX and lit-html:

```js
render() {
  let a = html`<div>${1}</div>`;
  let b = html`<div>${2}</div>`;
  return html`
    <div>${a}${b}</div>
 `;
}
```

This would produce two templates, vaguely like this:

```html
<template id="a"><div>{{}}</div></template>
<template id="b"><div>{{}}</div></template>
<template id="outer"><div>{{}}{{}}</div></template>
```

The node parts in `outer` need to be independently assignable to other Nodes.

**Comment by rniwa** [comment]
Creating two template parts is given. The question is whether any template parts should share a single text node or not.

**Comment by justinfagnani** [comment]
If they do share a Text node, then how would you set one part to a Node?

Maybe I'm misunderstanding the use for the Text node here. Are the Text nodes in question representing the Part, or just the anchors for the Part?

**Comment by rniwa** [comment]
Text nodes don't represent the anchor. It's the result of running the algorithm to apply node template part list. It doesn't really matter what the implementation.

The question is whether it's desirable for the final DOM to have two distinct `Text` nodes for each part, or they could share a a single `Text` node.
